[NCA-1683] Adding DRApp CLI tool (#851)
datarobot/datarobot-user-models: drapps/README.md

# DRApps

## About 

> DRApps is simple CLI tool for creating your custom application in Data Robot cloud.

Reviewer: ```suggestion
DRApps is simple CLI tool for creating your custom application in DataRobot.
```

> Custom application it small web server hosted by Data Robot in Kubernetes environment.
> Data Robot do not provide any storage resources for this server, but you can use Data
> Robot API and other Internet services.

Reviewer: The user is probably not concerned whether we run their app in kubernetes or not. That's an implementation detail we don't need to tell them about.

```suggestion
Custom applications does not provide any storage but you may access the full DataRobot API and other Internet services
```

## Installation

For using this CLI script you need only install python requirements with:

>     pip install -r requirements.txt

Reviewer: So to install this, the user need to clone the repository and then go to the cloned destination and install requirements? I think that's too much work and there's a simpler way for the user:

Can we make sure dependencies and the CLI is automatically installed when the user does

```
pip install datarobot-drum
```

This could put a `drapps` script i the user's `$PATH`.

Then the user could just use `drapps` from anywhere, right?

Author: In that case I need to make this script part of `datarobot_drum` package.
I though we keep this script as separate entity. At least for now. Am I wrong?

Reviewer: If it's hard to make that work, let's keep it as is right now.

But then you should also include a full example of cloning the code, creating a virtualenv and installing dependencies. The users are not going to be Python experts. They're data scientists. They will likely not know how to clone a github repository or install a virtualenv.

They're used to only working in notebooks

## Usage

    $ ./drapps.py --help
    Usage: drapps.py [OPTIONS]
    
      App that uses local file for create new custom application
    
    Options:
      -e, --base-env TEXT   Name or ID for execution environment  [required]
      -p, --path DIRECTORY  Path to folder with files that should be uploaded
                            [required]
      -n, --name TEXT       Name for new custom application. Default: CustomApp
      -t, --token TEXT      Pubic API access token.
      -E, --endpoint TEXT   Data Robot Public API endpoint
      --help                Show this message and exit.

### --base-env

UUID or name of Execution Environment that will be used as base for your app setup.
Execution Environment contains preinstalled essential libraries and packages, that 
can be used by your application.

You can find list of available Execution Environments in **Custom Model Workshop**
on [**Environments page**](https://app.datarobot.com/model-registry/custom-environments)

> 

Reviewer: Can we add something like:

> For a custom Streamlit application, use: ...

### --path

Path to folder which will be used to create custom application. Files from this folder
will be uploaded to Data Robot and used for creating custom application image. Custom
application will be started from this image.

> 

Reviewer: Can we add something like:

> To use the current working directory use `--path .`.

### --name

Name for your custom application. Default value is `CustomApp`. This name will be also

> used for generating name for custom application image by adding `Image` suffix.

Reviewer: Why does the user need to know what name their image has?

Author: There is no specific need in this info. But because customer have access to custom model image API, i wanted to make they know, what comes from where.

Also, you can provide token for the script through `DATAROBOT_API_TOKEN` environment
variable.

### --endpoint

URL to Public API of Data Robot. Default value: `https://app.datarobot.com/api/v2`.

You can provide URL to Public API through  `DATAROBOT_ENDPOINT` environment variable.

## Script work description

Script works in 4 steps:

1. Script uses `/api/v2/executionEnvironments/` endpoints for finding Execution
   Environment by name or UUID provided by user; checks if environment can be used
   for custom application and takes ID of latest version.

2. Script uses `/api/v2/customApplicationImages/` endpoint for finding or creating
   custom application image with name formed by adding `Image` suffix to application
   name.

3. Script uses `customApplicationImages/<appImageId>/versions` endpoints for creating
   new custom application image version, uploading to this version all files from
   folder provided by user and setting Execution Environment version found earlier.

4. Newly created custom application image version is used to starting a new application.

Result of script work you can find on [Applications](https://app.datarobot.com/applications)

> tab.

Reviewer: Why does the user need to know this?

I think this can be replaced with a new section on **Accessing your newly provisioned application**

Author: I just want to give more detailed description in case customer want to do something with API on their own.

But if you think it`s not the best idea, I can remove this part.

## Tips for your custom application

> Root folder of custom application should contain `start-app.sh` file, that will be

Reviewer: ```suggestion
Root folder of custom application must contain `start-app.sh` file, that will be
```

used as entry point for starting your application server.

> Web server of the application should listen on **8080** port.

Reviewer: ```suggestion
Web server of the application must listen on **8080** port.
```

# DRApps

## About 

DRApps is simple CLI tool for creating your custom application in DataRobot.

Custom applications does not provide any storage but you may access the full DataRobot API and other Internet services

## Installation

For using this CLI script you need only install python requirements with:

    pip install -r requirements.txt

## Usage

    $ ./drapps.py --help
    Usage: drapps.py [OPTIONS]
    
      App that uses local file for create new custom application
    
    Options:
      -e, --base-env TEXT   Name or ID for execution environment  [required]
      -p, --path DIRECTORY  Path to folder with files that should be uploaded
                            [required]
      -n, --name TEXT       Name for new custom application. Default: CustomApp
      -t, --token TEXT      Pubic API access token.

>       -E, --endpoint TEXT   Data Robot Public API endpoint

Reviewer 2: Cosmetic nit but still important IMO, anywhere that mentions `Data Robot` should be changed to `DataRobot`
